File: Homeworks/Homework4/Homework4.Tests/Parser.Tests.fs
module Homework4.Tests.Parser_Tests
open System
open Homework4
open Xunit

[<Fact>]
let ``ParseArguments_CorrectValues``() =
    Assert.Equal((12.,11.),Parser.parseArguments [|"12";"*";"11"|])
    Assert.Equal((213.,1234.),Parser.parseArguments [|"213";"x";"1234"|])
    
[<Fact>]
let ``ParseArguments_InvalidArgument1``() =
    let res() =
        let arg1 = Parser.parseArguments[|"x";"*";"11"|]
        printfn $"{arg1}"
    Assert.Throws<ArgumentException>(Action res)
    
[<Fact>]
let ``ParseArguments_InvalidArgument2``() =
    let res() =
        let arg1,arg2 = Parser.parseArguments[|"123";"*";"z"|]
        printfn $"{arg1}{arg2}"
    Assert.Throws<ArgumentException>(Action res)
    
[<Fact>]
let ``ParseToOperation_CorrectOperations``() =
    Assert.Equal(Operation.Plus,Parser.parseToOperation "+")
    Assert.Equal(Operation.Minus,Parser.parseToOperation "-")
    Assert.Equal(Operation.Multiply,Parser.parseToOperation "*")
    Assert.Equal(Operation.Divide,Parser.parseToOperation "/")
    
[<Fact>]
let ``ParseToOperation_UnknownOperation``() =
    Assert.Equal(Operation.Unknown,Parser.parseToOperation "x")

File: Homeworks/Homework2.IL/Homework2.Tests/Parser.Tests.cs
using Homework2.IL;
using Xunit;

namespace Homework2.Tests
{
    public class ParserTests
    {
        [Fact]
        public void TryParseArguments_Parse31Plus112_TrueReturned()
        {
            var res = Parser.TryParseArguments(new[] {"31","+","112"}, out var val1, out var val2);
            Assert.True(res);
            Assert.Equal(31,val1);
            Assert.Equal(112,val2);
        }

        [Fact]
        public void TryParseArguments_IncorrectOperation_CorrectParse_TrueReturned()
        {
            var res = Parser.TryParseArguments(new[] {"11", "x", "300"}, out var val1, out var val2);
            Assert.True(res);
            Assert.Equal(11,val1);
            Assert.Equal(300,val2);
        }

        [Fact]
        public void TryParseArguments_NotCalculateSyntax_FalseReturned()
        {
            var res =  Parser.TryParseArguments(new[] {"x", "+", "123"}, out var val1, out var val2);
            Assert.False(res);
        }
    }
}